src/coreclr/vm/reflectioninvocation.cpp

```diff
@@ -1321,49 +1298,17 @@ FCIMPL0(FC_BOOL_RET, ReflectionInvocation::TryEnsureSufficientExecutionStack)
 }
 FCIMPLEND
 
-FCIMPL4(void, ReflectionInvocation::MakeTypedReference, TypedByRef * value, Object* targetUNSAFE, ArrayBase* fldsUNSAFE, ReflectClassBaseObject *pFieldTypeUNSAFE)
+FCIMPL2(INT32, ReflectionInvocation::ComputeOffsetForTypedReference, INT32 fldsLen, FieldDesc** flds)
 {
-    CONTRACTL
-    {
-        FCALL_CHECK;
-        PRECONDITION(CheckPointer(targetUNSAFE));
-        PRECONDITION(CheckPointer(fldsUNSAFE));
-    }
-    CONTRACTL_END;
+    FCALL_CONTRACT;
 
     DWORD offset = 0;
-
-    struct _gc
+    for (INT32 i = 0; i < fldsLen; i++)
     {
-        OBJECTREF   target;
-        BASEARRAYREF flds;
-        REFLECTCLASSBASEREF refFieldType;
-    } gc;
-    gc.target  = (OBJECTREF)   targetUNSAFE;
-    gc.flds   = (BASEARRAYREF) fldsUNSAFE;
-    gc.refFieldType = (REFLECTCLASSBASEREF)ObjectToOBJECTREF(pFieldTypeUNSAFE);
-
-    TypeHandle fieldType = gc.refFieldType->GetType();
-
-    HELPER_METHOD_FRAME_BEGIN_PROTECT(gc);
-    GCPROTECT_BEGININTERIOR (value)
-
-    DWORD cnt = gc.flds->GetNumComponents();
-    FieldDesc** fields = (FieldDesc**)gc.flds->GetDataPtr();
-    for (DWORD i = 0; i < cnt; i++) {
-        FieldDesc* pField = fields[i];
+        FieldDesc* pField = flds[i];
         offset += pField->GetOffset();
     }
-
-        // Fields already are prohibted from having ArgIterator and RuntimeArgumentHandles
-    _ASSERTE(!gc.target->GetTypeHandle().GetMethodTable()->IsByRefLike());
-
-    // Create the ByRef
-    value->data = ((BYTE *)(gc.target->GetAddress() + offset)) + sizeof(Object);
-    value->type = fieldType;
-
-    GCPROTECT_END();
-    HELPER_METHOD_FRAME_END();
+    return (INT32)offset;
```

Comment:
1. What would it take to do this all in managed? it looks like FieldDesc::m_dwOffset is a bitfield, so it's probably not completely straightforward

2. I don't think this works for `EnCFieldDesc` (it didn't work before either, I'm pretty sure).  It's kind of the same flavor of problem as https://github.com/dotnet/runtime/issues/76702 was

Reply:
while I'm at it... what the heck is `FIELD_OFFSET_BIG_RVA` - is that only for static fields?  it looks like sometimes `GetOffset()` returns an RVA and not an instance offset?

Reply:
It would take a bit. The issue is large RVA scenarios where we fallback to metadata reading. Look for `FIELD_OFFSET_BIG_RVA`. Other than that, it would be possible. Not in this PR though.

Reply:
> I don't think this works for EnCFieldDesc (it didn't work before either, I'm pretty sure)

I agree that this does not work for EnC added fields. It would be nice to fix that while we are on it.

Also, the current implementation is working around historic limitations around `ref` by allocating array of IntPtr handles. I think it is unnecessary ceremony. I would be nice to streamline to implementation to something like this:

```csharp
public static TypedReference MakeTypedReference(object target, FieldInfo[] flds)
{
     ArgumentNullException.ThrowIfNull(target);
     ArgumentNullException.ThrowIfNull(flds);

     if (flds.Length == 0)
         throw new ArgumentException(SR.Arg_ArrayZeroError, nameof(flds));

      ref byte targetRef = Unsafe.NullRef<byte>();
      for (int i = 0; i < flds.Length; i++)
      {
           FieldInfo field = flds[i];

           .... existing argument validation ...

           if (i == 0)
           {
               targetByref = ref GetInstanceFieldAddress(target, field)
           }
           else
           {
               targetByref = ref GetValueTypeFieldAddres(ref targetRef, field);
           }
      }

      return new TypedReference(ref targetRef, targetType.Value);
}
```

Reply:
> I would be nice to streamline to implementation to something like this:

I'm fine doing that, but there are some issues with reconciling Mono and CoreCLR. The biggest difference is that Mono has a different layout for `TypedReference`.

https://github.com/dotnet/runtime/blob/b2c8aa27f89a9883eb3757303b580214cc84591f/src/mono/System.Private.CoreLib/src/System/TypedReference.Mono.cs#L11-L17

I didn't want to try and reconcile that in this PR since it was about removing HelperMethodFrames (HMF). The result is that it is largely the same logic, but can be improved. I personally would prefer to not try reconciling the two paths in this PR, but rather in a follow-up. The HMFs are removed and I _think_ the path forward to simplify and fix the EnC deficiency is clear.

Reply:
I would be fine with copy&pasting the existing implementation without any changes to Mono specific file and deal with it separately if necessary.

I am having hard time reviewing correctness of the converted implementation. For example, `GC.KeepAlive(flds); // Ensure FieldInfos are alive for passed in FieldDescs.` does not work` is needed. Is it trying to fix a bug in the existing implementation? If it is the case, it is not really 100% correct fix since the array of FieldInfos is owned by calling code that can be mutating it.

Reply:
> I am having hard time reviewing correctness of the converted implementation. 

The array is ensuring the `FieldInfo`s that we use to get at the associated `FieldDesc`s remain alive while we are using them. I agree it is not 100 % correct because, yes, the array contents could change.

> I would be fine with copy&pasting the existing implementation without any changes to Mono specific file and deal with it separately if necessary.

Okay, if this is acceptable, then I am happy to update the CoreCLR version to align with your suggestion involving `GetInstanceFieldAddress`/`GetValueTypeFieldAddress`. I think changing these names to be something akin to `GetRefToField` is more appropriate though. My push back is primarily on the word "Address" since we are getting back a `ref byte`, that is _not_ an address and I don't think we should treat it as such. Open to other suggestions.

Reply:
Just seeing this thread. Here is my coreclr version for reference:

```csharp
        public static TypedReference MakeTypedReference(object target, FieldInfo[] flds)
        {
            ArgumentNullException.ThrowIfNull(target);
            ArgumentNullException.ThrowIfNull(flds);

            if (flds.Length == 0)
                throw new ArgumentException(SR.Arg_ArrayZeroError, nameof(flds));

            int offset = 0;
            // For proper handling of Nullable<T> don't change GetType() to something like 'IsAssignableFrom'
            // Currently we can't make a TypedReference to fields of Nullable<T>, which is fine.
            RuntimeType targetType = (RuntimeType)target.GetType();
            for (int i = 0; i < flds.Length; i++)
            {
                RtFieldInfo? field = flds[i] as RtFieldInfo;
                if (field == null)
                    throw new ArgumentException(SR.Argument_MustBeRuntimeFieldInfo);

                if (field.IsStatic)
                    throw new ArgumentException(SR.Format(SR.Argument_TypedReferenceInvalidField, field.Name));

                if (targetType != field.GetDeclaringTypeInternal() && !targetType.IsSubclassOf(field.GetDeclaringTypeInternal()))
                    throw new MissingMemberException(SR.MissingMemberTypeRef);

                RuntimeType fieldType = (RuntimeType)field.FieldType;
                if (fieldType.IsPrimitive)
                    throw new ArgumentException(SR.Format(SR.Arg_TypeRefPrimitive, field.Name));

                if (i < (flds.Length - 1) && !fieldType.IsValueType)
                    throw new MissingMemberException(SR.MissingMemberNestErr);

                offset += RuntimeFieldHandle.GetInstanceFieldOffset(field);
                targetType = fieldType;
            }

            // Fields already are prohibted from having ArgIterator and RuntimeArgumentHandles
            Debug.Assert(!target.GetType().IsByRefLike); // Boxed value can never be ByRefLike

            // Create the ByRef
            return new TypedReference(ref Unsafe.Add(ref target.GetRawData(), offset), targetType.m_handle);
        }
```

Reply:
This is not correct with EnC added fields as discussed above. We need to be updating current byref instead of current offset as we are iterating over fields to make it work for EnC added fields.

Reply:
I agree that the helper method names in my snippet are bad. I do not have strong opinion on what the names should be. 

The prior art for naming method like this in public APIs is `GetArrayDataReference`, `GetPinnableReference`, If we like this naming pattern, it would lead to something like `GetFieldDataReference`.

Reply:
> The prior art for naming method like this in public APIs is GetArrayDataReference, GetPinnableReference, If we like this naming pattern, it would lead to something like GetFieldDataReference.

Sounds good.